src/core/content/util.ts

```diff
@@ -745,28 +745,70 @@ export function injectScriptIntoDocument(scriptUrl: string): void {
 	(document.head || document.documentElement).appendChild(script);
 }
 
+/**
+ * Handle async checks for DEBUG_LOGGING_ENABLED option while ensuring
+ * that logs are still printed in a predictable order.
+ */
+class DebugLogQueue {
+  private queue: {text: unknown, logType: DebugLogType}[] = [];
+  private isActive = false;
+
+  /**
+   * Enqueue a log message to be printed.
+   * @param text - Debug message
+   * @param logType - Log type
+   */
+  public push(text: unknown, logType: DebugLogType): void {
+    this.queue.push({text, logType});
+    this.start();
+  }
+
+  /**
+   * Process the queue to print logs in order.
+   */
+  private async start(): Promise<void> {
+    if (this.isActive) return;
+    this.isActive = true;
+
+    try {
+      for (let i = 0; i < 100 && this.queue.length > 0; i++) {
+        const currentMessage = this.queue.shift();
+        if (currentMessage && await this.shouldPrint()) {
+          const logFunc = console[currentMessage.logType];
+
+          if (typeof logFunc !== 'function') {
+            throw new TypeError(`Unknown log type: ${currentMessage.logType}`);
+          }
+
+          const message = `Web Scrobbler: ${currentMessage.text}`;
+          logFunc(message);
+        }
+      }
+      this.isActive = false;
+    } catch(err) {
+      this.isActive = false;
+    }
+  }
+
+  /**
+   * Check if the user has enabled debug logging in the extension options.
+   */
+  private async shouldPrint(): Promise<boolean> {
```

Comment:
I'm not entirely sure, but I'm leaning towards preferring this as a variable rather than function. So something like this:

```ts
private shouldPrint = Options.then((awaitedOptions) => {
		return Boolean(
			 awaitedOptions.getOption(awaitedOptions.DEBUG_LOGGING_ENABLED),
		);
});
```
Benefit is it's a lot faster and less demanding than having it as a function as it only actually fetches the option once.
Drawback of course being that if you change the option you will need to refresh the website for it to go through.

Reply:
To add to that, I'm not quite convinced the speed actually matters here, but it might on a slow HDD (and for modern machines there's the whole death by a thousand paper cuts thing)

Reply:
:+1: That seems pretty reasonable, users probably won't be toggling this option very often. I've switched to that here and it feels much nicer: edc6372